### DISCUSSION TITLE
getFolderTrackCount(index) prints file count of 1st folder regardless of index

### DISCUSSION THREAD

**Discussion by thre3eye** [Makuna/DFMiniMp3#142] [Q&A] [answered]
Thanks very much for this library - it's the most reliable one for me so far.
I am using it with an ESP32 C3 and a dfplayer HW-247 clone (GD3200B chip) and it's the only library that correctly prints file/folder count, volume, etc.

Except getFolderTrackCount only returns the number of` files in the 1st folder:

```
  uint16_t folderCount = dfmp3.getTotalFolderCount();
  Serial.printf("Total Folder #  :   %d \n", folderCount);
  for (uint16_t i = 0; i < folderCount; i++) {
    Serial.printf("Folder %d Track #:   %d \n", i, dfmp3.getFolderTrackCount(i));
  }
```
16:16:19.099 -> Total Folder #  :   5 
16:16:19.145 -> Folder 0 Track #:   3 
16:16:19.224 -> Folder 1 Track #:   3 
16:16:19.271 -> Folder 2 Track #:   3 
16:16:19.318 -> Folder 3 Track #:   3 
16:16:19.395 -> Folder 4 Track #:   3 

Any idea? Thanks!

**Comment by Makuna** [comment]
Put the following in your sketch before you include the library and rerun your test again.  The debug output should include low level packet dumps which will give me more information.
```
#define DfMiniMp3Debug 1
```
What are the names of your folders?  Some devices require the name to start with two digit folder number that is zero padded (it can end with anything) and some will just return the nth enumerated folder.
```
00Anything
01Nothing
02OneThing
```

**Comment by thre3eye** [comment]
Hi - thanks for following up.

I put our define in the 1st line of my .ino but am not seeing more output (in the Serial Monitor). Do I need to do other stuff? (sorry if it's obvious, while being a seasoned programmer I am quite new to Arduino).

My folders are simply named 01, 02, 03, ... 0n. The files are named 001_<xxx...>, 002_<xxx...>, 003_<xxx...>, .... 00n_<xxx...> in each folder. Folder 01 contains 3 files, 02 1 file, 03 18 files and 04 and 05 each 1 file, so the total is correct.

Thank you.

```
10:09:40.618 -> SD Card, online
10:09:40.618 -> Starting mp3...
10:09:43.507 -> Version         :   0 
10:09:43.554 -> Status source   :   0 
10:09:43.586 -> Status state    :   2 
10:09:43.679 -> Mode            :   5 
10:09:43.725 -> Volume          :   5 
10:09:43.803 -> Total Track #   :   24 
10:09:43.850 -> Total Folder #  :   5 
10:09:43.928 -> Folder 0 Track #:   3 
10:09:43.976 -> Folder 1 Track #:   3 
10:09:44.053 -> Folder 2 Track #:   3 
10:09:44.099 -> Folder 3 Track #:   3 
10:09:44.177 -> Folder 4 Track #:   3 
10:09:44.223 -> Current Track # :   1 
10:09:44.783 -> Play finished for #3
```

**Comment by Makuna** [comment]
If you look at [this entry in the Wiki](https://github.com/Makuna/DFMiniMp3/wiki/Debug-Features#dfminimp3debug), you can see what you should be seeing in the output.

I gave you the wrong value for the define, follow the wiki entry above.

**Comment by boerge1** [comment, marked as answer]
This is a known bug of the GD3200B chip. `getFolderTrackCount()` ignores the parameter folder and gives the track count of the folder last played from (on start is uses the first folder in the root directory).

We use the following workaround: 

```
#ifdef DFMiniMp3_T_CHIP_GD3200B
    uint8_t volume = mp3.getVolume();
    mp3.setVolume(0);
    delay(100);
    mp3.playFolderTrack(folder, 1);
    delay(500);
#endif
    trackCount = mp3.getFolderTrackCount(folder);
#ifdef DFMiniMp3_T_CHIP_GD3200B
    mp3.setVolume(volume);
#endif
```

**Comment by JonS9999** [comment]
This is my first post, and I may be way off base here, so feel free to tell
me to go be quiet and sit in the corner.  ;-)

If the purpose of the code snippet you provided is to determine the track
count, then it looks like if you are *not* using the GD3200B chip, the
audio file will start playing at regular volume.

Is this what you really meant?

#ifdef DFMiniMp3_T_CHIP_GD3200B
    uint8_t volume = mp3.getVolume()
    mp3.setVolume(0);
    delay(100);
    mp3.playFolderTrack(folder, 1);
    delay(500);
#endif
    trackCount = mp3.getFolderTrackCount(folder);
#ifdef DFMiniMp3_T_CHIP_GD3200B
    mp3.setVolume(volume);
#endif


Also, does the file automatically stop playing when the
getFolderTrackCount() call is made?

Again, please take these comments/observations with a grain of salt as I
may be way off base here.

Cheers.

-Jon

**Comment by thre3eye** [comment]
Ok, great. Thanks very much for the info. I have it working with the following code. It doesn't require delays and I am calling stop() so not to take chances.

```
  uint16_t folderCount = dfmp3.getTotalFolderCount();
  Serial.printf("Total Folder #  :   %d \n", folderCount);
  for (uint16_t i = 1; i <= folderCount; i++) {
    dfmp3.playFolderTrack(i, 1);
    dfmp3.stop();
    Serial.printf("Folder %d Track #:   %d \n", i, dfmp3.getFolderTrackCount(i));
  }
```

@Makuna here is also the output you requested.
Is this something you can/want to address, or should I close the issue with @boerge1 's answer?
```

20:27:18.697 -> Initializing mp3...
20:27:19.708 -> OUT  7e ff 06 0c 01 00 00 fe ee ef
20:27:19.755 -> IN  7e ff 06 40 00 00 03 fe b8 ef
20:27:19.755 -> OUT  7e ff 06 0c 01 00 00 fe ee ef
20:27:19.833 -> IN  7e ff 06 41 00 00 00 fe ba ef
20:27:20.602 -> IN  7e ff 06 3f 00 00 02 fe ba ef
20:27:20.602 -> SD Card, online
20:27:20.602 -> OUT  7e ff 06 09 01 00 02 fe ef ef
20:27:20.633 -> IN  7e ff 06 41 00 00 00 fe ba ef
20:27:20.633 -> OUT  7e ff 06 06 01 00 05 fe ef ef
20:27:20.727 -> IN  7e ff 06 41 00 00 00 fe ba ef
20:27:20.727 -> OUT  7e ff 06 07 01 00 04 fe ef ef
20:27:20.772 -> IN  7e ff 06 41 00 00 00 fe ba ef
20:27:20.772 -> OUT  7e ff 06 46 00 00 00 fe b5 ef
20:27:21.812 -> OUT  7e ff 06 46 00 00 00 fe b5 ef
20:27:22.831 -> OUT  7e ff 06 46 00 00 00 fe b5 ef
20:27:23.844 -> Version         :   0 
20:27:23.844 -> OUT  7e ff 06 42 00 00 00 fe b9 ef
20:27:23.892 -> IN  7e ff 06 42 00 00 02 fe b7 ef
20:27:23.892 -> Status source   :   0 
20:27:23.892 -> OUT  7e ff 06 42 00 00 00 fe b9 ef
20:27:23.970 -> IN  7e ff 06 42 00 00 02 fe b7 ef
20:27:23.970 -> Status state    :   2 
20:27:23.970 -> OUT  7e ff 06 45 00 00 00 fe b6 ef
20:27:24.013 -> IN  7e ff 06 45 00 00 05 fe b1 ef
20:27:24.013 -> Mode            :   5 
20:27:24.013 -> OUT  7e ff 06 43 00 00 00 fe b8 ef
20:27:24.090 -> IN  7e ff 06 43 00 00 05 fe b3 ef
20:27:24.090 -> Volume          :   5 
20:27:24.090 -> OUT  7e ff 06 48 00 00 00 fe b3 ef
20:27:24.135 -> IN  7e ff 06 48 00 00 18 fe 9b ef
20:27:24.135 -> Total Track #   :   24 
20:27:24.135 -> OUT  7e ff 06 4f 00 00 00 fe ac ef
20:27:24.212 -> IN  7e ff 06 4f 00 00 05 fe a7 ef
20:27:24.212 -> Total Folder #  :   5 
20:27:24.212 -> OUT  7e ff 06 4e 00 00 00 fe ad ef
20:27:24.259 -> IN  7e ff 06 4e 00 00 03 fe aa ef
20:27:24.259 -> Folder 0 Track #:   3 
20:27:24.259 -> OUT  7e ff 06 4e 00 00 01 fe ac ef
20:27:24.338 -> IN  7e ff 06 4e 00 00 03 fe aa ef
20:27:24.338 -> Folder 1 Track #:   3 
20:27:24.338 -> OUT  7e ff 06 4e 00 00 02 fe ab ef
20:27:24.385 -> IN  7e ff 06 4e 00 00 03 fe aa ef
20:27:24.385 -> Folder 2 Track #:   3 
20:27:24.385 -> OUT  7e ff 06 4e 00 00 03 fe aa ef
20:27:24.462 -> IN  7e ff 06 4e 00 00 03 fe aa ef
20:27:24.462 -> Folder 3 Track #:   3 
20:27:24.462 -> OUT  7e ff 06 4e 00 00 04 fe a9 ef
20:27:24.510 -> IN  7e ff 06 4e 00 00 03 fe aa ef
20:27:24.510 -> Folder 4 Track #:   3 
20:27:24.510 -> OUT  7e ff 06 4c 00 00 00 fe af ef
20:27:24.541 -> IN  7e ff 06 4c 00 00 01 fe ae ef
20:27:24.541 -> Current Track # :   1 
20:27:24.541 -> OUT  7e ff 06 0f 01 01 01 fe e9 ef
20:27:24.635 -> IN  7e ff 06 41 00 00 00 fe ba ef

```

**Reply by Makuna** [reply]
This output confirms what was stated, the device is returning the same number no matter what folder number you gave it.  
It is interesting to note that it doesn't support the software version query, it just ignores it, no error, no return.

So, yeah, closed.  I updated the Wiki for this Method to point to this discussion with the NOTE about the issue.

I created an issue to track hiding the issue from the sketch writer; not a high priority, but maybe in the future I might do something about it.
https://github.com/Makuna/DFMiniMp3/issues/145
